### DISCUSSION TITLE
Ideas for local glyph rasterization

### DISCUSSION THREAD

**Discussion by wipfli** [maplibre/maplibre-gl-js#3111] [Ideas]
@bdon's pull request on doubling the resolution of locally rasterized glyphs #3006 made me wonder how else we could use the approach of generating SDFs in the client rather than on a server. This discussion is to collect some ideas...

1. We could use local glyphs for all letters, so not only CJK but also Latin, Greek, etc. This would mean people don't have to generate the SDF fonts ahead of time but rather ship a .ttf font to the client or rely on system fonts for when letters are missing.
1. Once we rasterize glyphs in the client, the performance impact of larger resolutions for larger letters becomes has different constraints. Previously, the problem was download size of large letters and gpu memory. Now it would be only gpu memory and compute time in the client. Maybe big letters with sharp corners could become feasible... see https://github.com/wipfli/double-resolution-cjk-maplibre#quadruple-resolution
1. Users could use webfonts directly in maplibre gl js such as roboto.ttf without having to convert them ahead of time.
1. Users could use more of the normal CSS font properties like font weight for example.
1. Last but not least, server-side text segmentation together with local segment rasterization could be an interesting path for indic script support. See https://github.com/wipfli/index-by-grapheme

**Comment by bdon** [comment]
If you ship a TTF font to the client, then MapLibre needs to build a mechanism to block on the browser Fonts API until it knows that font is successfully loaded. 

I did such a thing here: https://github.com/protomaps/protomaps-leaflet/blob/main/src/task.ts#L7
Inspired by here: https://github.com/tangrams/tangram/blob/master/src/styles/text/font_manager.js
